### ISSUE TITLE
runner register - failing to match correct name

### ISSUE THREAD

**Issue by schewara** [voxpupuli/puppet-gitlab_ci_runner#45]
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 5.5.8
- Ruby: 2.3.3
- Distribution: debian/stretch
- Module version: master

## How to reproduce (e.g Puppet code you use)

```ruby
gitlab_ci_runner::runners:
  "shared":
    name: "shared_%{facts.hostname}"
    docker-image: "debian:stretch"
```
## What are you seeing

subsequent Puppet runs keep registering new runners

## What behaviour did you expect instead

no change

## Output log

```
Debug: Exec[Register_runner_shared](provider=posix): Executing check '/bin/grep '"shared_runner1"' /etc/gitlab-runner/config.toml'
Debug: Executing: '/bin/grep '"shared_runner1"' /etc/gitlab-runner/config.toml'
Debug: Exec[Register_runner_shared](provider=posix): Executing '/usr/bin/gitlab-runner register -n --name=shared-runner1 --paused --registration-token=abcdefg --executor=docker --url=https://git.my.tld --docker-image=debian:stretch --docker-disable-entrypoint-overwrite=false --docker-oom-kill-disable=false --docker-disable-cache=false --docker-shm-size=0 --docker-privileged=false --docker-tlsverify=false --docker-volumes=/cache'
Debug: Executing: '/usr/bin/gitlab-runner register -n --name=shared-runner1 --paused --registration-token=abcdefg --executor=docker --url=https://git.my.tld --docker-image=debian:stretch --docker-disable-entrypoint-overwrite=false --docker-oom-kill-disable=false --docker-disable-cache=false --docker-shm-size=0 --docker-privileged=false --docker-tlsverify=false --docker-volumes=/cache'
Notice: /Stage[main]/Gitlab_ci_runner/Gitlab_ci_runner::Runner[shared]/Exec[Register_runner_shared]/returns: executed successfully
Debug: /Stage[main]/Gitlab_ci_runner/Gitlab_ci_runner::Runner[shared]/Exec[Register_runner_shared]: The container Gitlab_ci_runner::Runner[shared] will propagate my refresh event
Debug: Gitlab_ci_runner::Runner[shared]: The container Class[Gitlab_ci_runner] will propagate my refresh event
```
## Any additional information you'd like to impart

`$parameters_array_no_underscores = regsubst($parameters_array, '_', '-', 'G')`
changes the `_` to `-` in the name, which causes the grep to fail, as it is looking for the original value `$runner_name = $_config['name']`

Workaround for now is to adjust the names with dashes instead of underscores.

**Comment by EdgeJ** [comment]
Rather than start a new issue, I'll comment here.  A similar behavior with the same root cause is that the `regsubst` method call also causes registration using a token with an underscore in it to fail, by converting the underscore to a hyphen.

E.g. if you attempt to register with a token like 'Aze2_fg', the Puppet run fails with an error:
```
Error: '/usr/bin/gitlab-runner register -n --url=https://gitlab.example.com --registration-token=Aze2-fg' --name=gitlab-runner --executor=shell' returned 1 instead of one of [0]
```

**Comment by baurmatt** [comment]
Hey, I've fixed this in #74 :)

**Comment by smortex** [comment]
The issues and workarounds listed here where part of code which was removed with version 4.0.0 of the module.  They are hopefully fixed, so closing.
